sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryCredentialsPolicy.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Containers.ContainerRegistry
+{
+    /// <summary>
+    /// Challenge-based authentication policy for Container Registry Service.
+    ///
+    /// The challenge-based authorization flow for ACR is illustrated in the following steps.
+    /// For example, GET /api/v1/acr/repositories translates into the following calls.
+    ///
+    /// Step 1: GET /api/v1/acr/repositories
+    /// Return Header: 401: www-authenticate header - Bearer realm="{url}",service="{serviceName}",scope="{scope}",error="invalid_token"
+    ///
+    /// Step 2: Retrieve the serviceName, scope from the WWW-Authenticate header.  (Parse the string.)
+    ///
+    /// Step 3: POST /api/oauth2/exchange
+    /// Request Body : { service, scope, grant-type, aadToken with ARM scope }
+    /// Response Body: { acrRefreshToken }
+    ///
+    /// Step 4: POST /api/oauth2/token
+    /// Request Body: { acrRefreshToken, scope, grant-type }
+    /// Response Body: { acrAccessToken }
+    ///
+    /// Step 5: GET /api/v1/acr/repositories
+    /// Request Header: { Bearer acrTokenAccess }
+    /// </summary>
+    internal class ContainerRegistryCredentialsPolicy : BearerTokenChallengeAuthenticationPolicy
```

Comment:
Do we have tests for this policy?

Reply:
Yes! see: sdk/containerregistry/Azure.Containers.ContainerRegistry/tests/Authentication/ContainerRegistryChallengeAuthenticationPolicyTest.cs